### PR TITLE
Jax tracing fix

### PR DESCRIPTION
**Pull request by nicolaspi**
Fixes https://github.com/keras-team/keras/issues/20402
Fixes https://github.com/keras-team/keras/issues/20411

The solution uses an approach similar to TensorFlow with `EpochIterator` returning an iterator instead of a list of batches. This prevents data loading bottlenecks when `steps_per_execution` > 1. It is enough to have a non-jitted for loop that iterates over the execution steps and makes calls to the inner `train/test/predict_step` function thanks to Jax's asynchronous dispatch.

This [benchmark](https://gist.github.com/nicolaspi/6c6483abffb57dec3db79a6e4c1efa1f) shows constant compilation time and memory usage with an improvement for prediction speed:

Before the changes:
```
steps_per_execution = 1, model retracing count: 1, memory usage: 887MB, compilation overhead 1.25 seconds, inference speed: 0.284 ms/sample.
steps_per_execution = 5, model retracing count: 5, memory usage: 1638MB, compilation overhead 3.88 seconds, inference speed: 0.310 ms/sample.
steps_per_execution = 10, model retracing count: 10, memory usage: 3211MB, compilation overhead 7.32 seconds, inference speed: 0.377 ms/sample.
```

After the changes:
```
steps_per_execution = 1, model retracing count: 1, memory usage: 887MB, compilation overhead 1.38 seconds, inference speed: 0.281 ms/sample.
steps_per_execution = 5, model retracing count: 1, memory usage: 887MB, compilation overhead 1.96 seconds, inference speed: 0.277 ms/sample.
steps_per_execution = 10, model retracing count: 1, memory usage: 887MB, compilation overhead 2.07 seconds, inference speed: 0.273 ms/sample.
steps_per_execution = 50, model retracing count: 1, memory usage: 887MB, compilation overhead 2.01 seconds, inference speed: 0.276 ms/sample.
steps_per_execution = 100, model retracing count: 1, memory usage: 887MB, compilation overhead 0.81 seconds, inference speed: 0.276 ms/sample.
```

This [benchmark](https://colab.research.google.com/gist/nicolaspi/ef2c37efb7f9c9bbcdf2fcfbb6d45ae6/untitled5.ipynb#scrollTo=gzRyT4VrhOJW) for training also shows an improvement in speed:

Before the changes (`steps_per_execution=1`; higher values gave slower training):
```
Epoch 1/40
160/160 ━━━━━━━━━━━━━━━━━━━━ 49s 143ms/step - accuracy: 0.0101 - loss: 4.9874 - val_accuracy: 0.0081 - val_loss: 4.7888
Epoch 2/40
160/160 ━━━━━━━━━━━━━━━━━━━━ 20s 86ms/step - accuracy: 0.0260 - loss: 4.6006 - val_accuracy: 0.0083 - val_loss: 4.8747
Epoch 3/40
160/160 ━━━━━━━━━━━━━━━━━━━━ 14s 86ms/step - accuracy: 0.0392 - loss: 4.3919 - val_accuracy: 0.0083 - val_loss: 5.1281
```

After the changes (`steps_per_execution=32`):
```
Epoch 1/40
160/160 ━━━━━━━━━━━━━━━━━━━━ 49s 90ms/step - accuracy: 0.0122 - loss: 4.9216 - val_accuracy: 0.0095 - val_loss: 4.7878
Epoch 2/40
160/160 ━━━━━━━━━━━━━━━━━━━━ 20s 72ms/step - accuracy: 0.0298 - loss: 4.5174 - val_accuracy: 0.0094 - val_loss: 4.9087
Epoch 3/40
160/160 ━━━━━━━━━━━━━━━━━━━━ 14s 72ms/step - accuracy: 0.0423 - loss: 4.3515 - val_accuracy: 0.0094 - val_loss: 5.2104
```

### CODE REVIEW

File: keras/src/backend/jax/trainer.py [hunk 1]
```diff
@@ -224,108 +223,109 @@ def predict_step(self, state, data):
...
+                            lambda t1, t2: jax.numpy.concatenate([t1, t2]),
+                            output,
+                            next_output,
+                        )
+                    return output
+
+                if not self.run_eagerly and self.jit_compile:
+                    concatenate = jax.jit(concatenate)
+
+                def function(state, iterator):
```
Comment (fchollet) [inline keras/src/backend/jax/trainer.py:243]: Please use an explicit name, like `iterator_step` or something like that

File: keras/src/backend/jax/trainer.py [hunk 2]
```diff
@@ -224,108 +223,109 @@ def predict_step(self, state, data):
...
-        def one_train_step(state, data):
-            data = data[0]
-            return self.train_step(state, data)
+            else:
 
-        def multi_train_steps(state, data):
-            for single_step_data in data:
-                logs, state = one_train_step(state, [single_step_data])
-            return logs, state
+                def function(state, iterator):
```
Comment (fchollet) [inline keras/src/backend/jax/trainer.py:259]: Same here

File: keras/src/trainers/data_adapters/generator_data_adapter.py [hunk 3]
```diff
@@ -49,7 +49,7 @@ def convert_to_tf(x, spec):
             return x
 
         def get_tf_iterator():
-            for batch in self.generator:
+            for batch in self.generator():
```
Comment (fchollet) [inline keras/src/trainers/data_adapters/generator_data_adapter.py:52]: Why is this necessary -- it's `generator` iterable?

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: keras/src/backend/jax/trainer.py (hunk 1 above)
Comment (nicolaspi) [reply keras/src/backend/jax/trainer.py:243]: Done

File: keras/src/backend/jax/trainer.py (hunk 2 above)
Comment (nicolaspi) [reply keras/src/backend/jax/trainer.py:259]: Done

File: keras/src/trainers/data_adapters/generator_data_adapter.py (hunk 3 above)
Comment (nicolaspi) [reply keras/src/trainers/data_adapters/generator_data_adapter.py:52]: This is for fixing https://github.com/keras-team/keras/issues/20402
We need to embed `generator` into a lambda to make it able to reset:
https://github.com/keras-team/keras/blob/fe397ed1af44670a838135a9804fab616f5ed4b8/keras/src/trainers/data_adapters/generator_data_adapter.py#L87

**Review by fchollet — APPROVED** [review]
LGTM, thank you for the contribution!
